tol_lab_share/messages/mlwh/create_aliquot_message.py

```diff
@@ -0,0 +1,134 @@
+import json
+import logging
+from typing import Any, Callable
+
+from lab_share_lib.rabbit.basic_publisher import BasicPublisher
+
+from tol_lab_share.constants import RABBITMQ_ROUTING_KEY_CREATE_ALIQUOT
+from tol_lab_share.error_codes import ErrorCode
+from tol_lab_share.helpers import get_config
+from tol_lab_share.messages.properties import MessageProperty
+from tol_lab_share.messages.properties.simple import Value
+
+logger = logging.getLogger(__name__)
+
+
+class Aliquot:
+    """Class that serialises into an aliquot"""
+
+    def __init__(self):
+        """Constructor that initializes an aliquot message"""
+        self.id_lims: str | None = None
+        self.lims_uuid: str | None = None
+        self.aliquot_type: str | None = None
+        self.source_type: str | None = None
+        self.source_barcode: str | None = None
+        self.sample_name: str | None = None
+        self.used_by_type: str | None = None
+        self.used_by_barcode: str | None = None
+        self.volume: float | None = None
+        self.concentration: float | None = None
+        self.insert_size: int | None = None
+        self.last_updated: str | None = None
+        self.recorded_at: str | None = None
+        self.created_at: str | None = None
+
+    def to_dict(self) -> dict:
+        """Convert the aliquot message to a JSON string"""
+        return {
+            "id_lims": self.id_lims,
+            "lims_uuid": str(self.lims_uuid),
+            "aliquot_type": self.aliquot_type,
+            "source_type": self.source_type,
+            "source_barcode": self.source_barcode,
+            "sample_name": self.sample_name,
+            "used_by_type": self.used_by_type,
+            "used_by_barcode": self.used_by_barcode,
+            "volume": self.volume,
+            "concentration": self.concentration,
+            "insert_size": self.insert_size,
+            "last_updated": self.last_updated,
+            "recorded_at": self.recorded_at,
+            "created_at": self.created_at,
+        }
+
+
+class CreateAliquotInWarehouseMessage(MessageProperty):
```

Comment:
I understand why you followed the pattern with the `MessageProperty` class, but I really think this is a pattern I wanted to step away from.  We shouldn't be doing such extensive validation of the values in the message when we're in full control of the message creation.  In almost every case, Avro schemas provide us with all the validation we need that values are of the correct type.  I appreciate that we might not have a schema defined for the message going into the MLWH RabbitMQ, but perhaps we should have?

Reply:
I actually agree, as I had the same thoughts on the validations; I realised it when I was writing the tests.

Reply:
And I suppose it's not _that_ difficult to get the schema in and do the validations at least on the producer side (`tol-lab-share` in this case). I'll put this into the conversation tomorrow.

Reply:
When I say using Avro, I do only mean on the publisher here.  We needn't modify how MLWH works as long as it can support the JSON encoded format of Avro.  If that's not the case, don't bother going down this route and do traditional type validation.  JSON format isn't highly supported across all libraries for Avro, but `fastavro` as used here is very good at it.

Reply:
mm well, the warehouse is a thin Rails application (there's technical term for such Rails applications; I can't remember it!) that has the queue as some sort of a sidecar so if it supports JSON formatting I guess there's no harm in using it. And I think we have integrating schema validations at the warehouse side in our roadmap if I remember our discussions before correctly.

Reply:
@sdjmchattie well, maybe I am looking in the wrong documentation, but [`avro` gem](https://www.rubydoc.info/gems/avro/1.8.1) in Ruby does support binary encoding and decoding. It doesn't say anything about dealing with a JSON-encoded format. For instance, when I JSON-encode (with [this schema](https://redpanda.uat.psd.sanger.ac.uk/console/schema-registry/subjects/create-aliquot-in-mlwh?version=latest)) the following message using `fastavro`:

```json
{
  "messageUuid": "b66d5daa-f1d5-4a74-9e93-969971043a42",
  "messageCreateDateUtc": 1578836705000,
  "limsId": "LIMS123",
  "limsUuid": "a8e4a3c3-954a-445b-b92f-bcec84029a3d",
  "aliquotType": "derived",
  "sourceType": "sample",
  "sourceBarcode": "SRC123456789",
  "sampleName": "Sample123",
  "usedByType": "library",
  "usedByBarcode": "LIB123456789",
  "volume": 150.5,
  "concentration": 20.0,
  "insertSize": 500,
  "recordedAt": 1578836705000
}
```

I get the following:

```json
{
  "messageUuid": "b66d5daa-f1d5-4a74-9e93-969971043a42",
  "messageCreateDateUtc": 1578836705000,
  "limsId": "LIMS123",
  "limsUuid": "a8e4a3c3-954a-445b-b92f-bcec84029a3d",
  "aliquotType": "derived",
  "sourceType": "sample",
  "sourceBarcode": "SRC123456789",
  "sampleName": "Sample123",
  "usedByType": "library",
  "usedByBarcode": "LIB123456789",
  "volume": 150.5,
  "concentration": {
    "float": 20
  },
  "insertSize": {
    "int": 500
  },
  "recordedAt": 1578836705000
}
```

I guess `avro` gem requires the ability to decode a message like above if we are going to json-encode it in tls? The warehouse expects the message to be of the structure similar to the message I encoded.

Reply:
The Ruby gem does not support JSON encoding or decoding.  My assumption is that the message on the MLWH side will not be changed and we won't be using Avro there.  I see that union types show the type in the JSON format for Avro.  I didn't know that and I don't think this will allow it to be suitable for creating the MLWH message.